src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+/*-------------------------------------------------------------------------
+ *
+ * shard_pruning.c
+ *   Shard pruning related code.
+ *
+ * Shard pruning works in two phases: First it analyses the expressions that
+ * are pruned upon, and builds a list of ORed constraints for pruning. This
+ * happens in PrunableExpressions().  The reason for keeping this separate
+ * from actually pruning away shards is that it has to only be done once for
+ * all shards.
+ *
+ * In the second phase, the shards are pruned using the constraints built in
+ * the first phase.  Each constraint (a set of ANDed constraints) is evaluated
+ * by PruneOne(), and the results of all results is then ORed/unioned
+ * together.  Depending on the types of constraints found, and the type of
+ * partitioning used, different methods of pruning are employed.  Ranging from
+ * hash-lookups (cf. FastShardPruning), binary search for equality constraints
+ * (also FastShardPruning), more complex binary search for inequality
+ * constraints (PruneWithBoundaries), to just verifying the constraint on each
+ * shard (ExhaustivePrune).
+ *
+ * Copyright (c) 2014-2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "distributed/shard_pruning.h"
+
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "catalog/pg_collation.h"
+#include "catalog/pg_type.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_planner.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/shardinterval_utils.h"
+#include "distributed/pg_dist_partition.h"
+#include "distributed/worker_protocol.h"
+#include "nodes/nodeFuncs.h"
+#include "optimizer/clauses.h"
+#include "utils/catcache.h"
+#include "utils/lsyscache.h"
+#include "utils/memutils.h"
+
+/*
+ * A pruning instance is a set of ANDed constraints on a partition key.
+ */
+typedef struct PruningInstance
+{
+	/*
+	 * This constraint never evaluates to false, i.e. pruning does not have to
+	 * be performed.
+	 */
+	bool alwaysFalse;
+
+	/*
+	 * Constraints on the partition column value.  If multiple values are
+	 * found the more restrictive one should be stored here.  Even in case of
+	 * a hash-partitioned table, actual column-values are stored here, *not*
+	 * hashed values.
+	 */
+	Const *lessConsts;
+	Const *lessEqualConsts;
+	Const *equalConsts;
+	Const *greaterEqualConsts;
+	Const *greaterConsts;
+
+	/*
+	 * Constraint using a pre-hashed column value. The constant will store the
+	 * hashed value, not the original value of the restriction.
+	 */
+	Const *hashedEqualConsts;
+
+	/*
+	 * Types of constraints not understood.  We could theoretically try more
+	 * expensive methods of pruning if any such restrictions are found.
+	 *
+	 * TODO: any actual use for this? Right now there seems little point.
```

Comment:
Right - I was more thinking whether we should keep the field for the moment, given it's not used atm.